Bars bar component refactoring

mattlangeman opened the pull request:
This replaces the previous PR #45.

The goal was refactoring Bars.svelte to better handle bar highlights.

I went with keeping the two different props (bar and area) for the Highlight component. This made more sense to me when dealing with examples like the multi-bar example for the Tooltip component.

I was not able to use setContext and getContext because Highlight is a sibling of Bars, not a child. I initially thought it would get the context from the closest sibling, but that is not the case. Therefore if you choose to have non-default values for thins like stroke, radius, and strokeWidth need to be passed in with the bar prop.

Open to feedback on improving this PR.

techniq commented:
Thanks Matt, looking good.  I just skimmed through it via Github and Vercel preview, but looks great!

A few suggestions:

- I think I'd like to keep this example as is (with just the Highlight rect), but maybe remove/replace the "with Tooltip and Bar only Highlight" and have this be "with Tooltip and Bar Highlight"

![image](https://github.com/techniq/layerchart/assets/177476/d02949b7-5290-40d4-9996-a818c9532b11)

I also think we could collapse
```svelte
<Highlight
  bar={{
    class: "fill-accent-500",
    stroke: "black",
    strokeWidth: 1,
    radius: 4,
  }}
/>
<Highlight area />
```
to just 
```svelte
<Highlight
  area
  bar={{
    class: "fill-accent-500",
    stroke: "black",
    strokeWidth: 1,
    radius: 4,
  }}
/>
```
not sure if you ran into a stack issue, which I figure we could just swap the render order them them if so.

- It might also be nice if for this example the bars were purple until you hovered the chart, then just the active one is purple, and the others are grey.  You should be able to do this with tailwind's `group-hover:`.   I have an example of this, but I need to find it...  I could also make that change later after merging the PR

- Maybe add the same example for Horizontal Bars 

- Side note, we should be able to reproduce the npm weekly download Sparkline chart once I allow better tooltip styling overrides (but the interaction and placement is all supported now).


https://github.com/techniq/layerchart/assets/177476/5cb7ab6e-f07e-4850-9e71-7350353faaa1

mattlangeman commented:
>     * I think I'd like to keep this example as is (with just the Highlight rect), but maybe remove/replace the "with Tooltip and Bar only Highlight" and have this be "with Tooltip and Bar Highlight"

I changed back to original "with Tooltip and Highlight" and made the "with Tooltip and Bar Highlight" show both area and bar.
 
> I also think we could collapse
> 
> ```svelte
> <Highlight
>   bar={{
>     class: "fill-accent-500",
>     stroke: "black",
>     strokeWidth: 1,
>     radius: 4,
>   }}
> />
> <Highlight area />
> ```
> 
> to just
> 
> ```svelte
> <Highlight
>   area
>   bar={{
>     class: "fill-accent-500",
>     stroke: "black",
>     strokeWidth: 1,
>     radius: 4,
>   }}
> />
> ```
> 
> not sure if you ran into a stack issue, which I figure we could just swap the render order them them if so.

Yes, the collapsing works fine. Highlight renders area highlight before bar highlight.

>     * It might also be nice if for this example the bars were purple until you hovered the chart, then just the active one is purple, and the others are grey.  You should be able to do this with tailwind's `group-hover:`.   I have an example of this, but I need to find it...  I could also make that change later after merging the PR

I gave this a very quick try, but it didn't work as I had it, so I'll leave this one to you for now.

>     * Maybe add the same example for Horizontal Bars

I added a "with Tooltip and Bar Highlight" to the horizontal bar examples.
> 
>     * Side note, we should be able to reproduce the npm weekly download Sparkline chart once I allow better tooltip styling overrides (but the interaction and placement is all supported now).

Excellent!

techniq reviewed (approved):
Looks great, thanks!

techniq commented:
@mattlangeman I [updated](https://github.com/techniq/layerchart/commit/ed042942aff48ffe5d44b3893378b2fbe1178d95) the examples with the gray on chart hover change... 

https://github.com/techniq/layerchart/assets/177476/a1715cfe-52e1-4080-af9e-9e8762cbeda9

I'll work on pushing a new release tonight (or at least this weekend).  I have a few things I might try to knock out before I do.

Thanks for this PR, I think it looks great!

mattlangeman commented:
> I'll work on pushing a new release tonight (or at least this weekend). I have a few things I might try to knock out before I do.

Sounds good. I just added  issue #54 related to the sort order of these date series examples. Looks to be sorting by alphabetical day of the week.
